grafana_client/client.py

```diff
@@ -119,72 +126,145 @@ def construct_api_url():
+            if isinstance(self.auth, niquests.auth.AuthBase):
                 pass
             elif isinstance(self.auth, tuple):
-                self.auth = requests.auth.HTTPBasicAuth(*self.auth)
+                self.auth = niquests.auth.HTTPBasicAuth(*self.auth)
             else:
                 self.auth = TokenAuth(self.auth)
 
+    def _make_url(self, url):
+        return f"{self.url}{url}"
+
+    @staticmethod
+    def _ensure_valid_json_arg(json):
+        if json is not None and not isinstance(json, (dict, list)):
+            raise TypeError(
+                f"JSON request payload has invalid shape. "
+                f"Accepted are dictionaries and lists. "
+                f"The type is: {type(json)}"
+            )
+
+    @staticmethod
+    def _extract_from_response(r, accept_empty_json):
+        if r.status_code >= 400:
+            try:
+                response = r.json()
+            except ValueError:
+                response = r.text
+            message = response["message"] if isinstance(response, dict) and "message" in response else r.text
+
+            if 500 <= r.status_code < 600:
+                raise GrafanaServerError(
+                    r.status_code,
+                    response,
+                    "Server Error {0}: {1}".format(r.status_code, message),
+                )
+            elif r.status_code == 400:
+                raise GrafanaBadInputError(response)
+            elif r.status_code == 401:
+                raise GrafanaUnauthorizedError(response)
+            elif 400 <= r.status_code < 500:
+                raise GrafanaClientError(
+                    r.status_code,
+                    response,
+                    "Client Error {0}: {1}".format(r.status_code, message),
+                )
+
+        # `204 No Content` responses have an empty response body,
+        # so it doesn't decode well from JSON.
+        if r.status_code == 204:
+            return None
+
+        # The "Tempo" data source responds with text/plain.
+        content_type = r.headers.get("Content-Type", "")
+        if content_type.startswith("text/"):
+            return r.text
+        try:
+            return r.json()
+        except JSONDecodeError:
+            if accept_empty_json and r.text == "":
+                return ""
+            else:
+                raise
+
     def __getattr__(self, item):
         def __request_runner(url, json=None, data=None, headers=None, accept_empty_json=False):
-            __url = "%s%s" % (self.url, url)
+            __url = self._make_url(url)
             # Sanity checks.
-            if json is not None and not isinstance(json, (dict, list)):
-                raise TypeError(
-                    f"JSON request payload has invalid shape. "
-                    f"Accepted are dictionaries and lists. "
-                    f"The type is: {type(json)}"
+            GrafanaClient._ensure_valid_json_arg(json)
```

Comment:
Could still use `self` here, like `self._make_url()`, no?
```suggestion
            self._ensure_valid_json_arg(json)
```

Reply:
You can, Python allow it. Let's do that.